crypto/pkcs11key/key_test.go

```diff
@@ -84,47 +257,96 @@ func (c mockCtx) SignInit(sh pkcs11.SessionHandle, m []*pkcs11.Mechanism, o pkcs
 }
 
 func (c mockCtx) Sign(sh pkcs11.SessionHandle, message []byte) ([]byte, error) {
-	return []byte("some signed data"), nil
+	return message, nil
 }
 
-func TestSetup(t *testing.T) {
-	ps := &Key{
-		module:     mockCtx{},
-		tokenLabel: "token label",
-		pin:        "unused",
-	}
-	err := ps.setup("private key label")
-	if err != nil {
-		t.Errorf("Failed to set up Key: %s", err)
-	}
-}
-
-func setup(t *testing.T) *Key {
+func setup(t *testing.T, label string) *Key {
 	ps := Key{
 		module:     mockCtx{},
 		tokenLabel: "token label",
 		pin:        "unused",
 	}
-	err := ps.setup("private key label")
+	err := ps.setup(label)
 	if err != nil {
 		t.Fatalf("Failed to set up Key: %s", err)
 	}
 	return &ps
 }
 
-func TestSign(t *testing.T) {
-	ps := setup(t)
+var signInput = []byte("1234567890 1234567890 1234567890")
+
+func sign(t *testing.T, ps *Key) []byte {
 	// Sign input must be exactly 32 bytes to match SHA256 size. In normally
 	// usage, Sign would be called by e.g. x509.CreateCertificate, which would
 	// handle padding to the necessary size.
-	signInput := []byte("1234567890 1234567890 1234567890")
 	output, err := ps.Sign(rand.Reader, signInput, crypto.SHA256)
 	if err != nil {
 		t.Fatalf("Failed to sign: %s", err)
 	}
-	if !bytes.Equal(output, []byte("some signed data")) {
+
+	if len(output) < len(signInput) {
+		t.Fatalf("Invalid signature size")
+	}
+
+	i := len(output) - len(signInput)
+	if !bytes.Equal(output[i:], signInput) {
 		t.Fatal("Incorrect sign output")
 	}
+	return output
+}
+
+func TestSign(t *testing.T) {
+	ps := setup(t, "rsa")
+	sig := sign(t, ps)
+
+	// Check that the RSA signature starts with the SHA256 hash prefix
+	var sha256Pre = []byte{0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86,
+		0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20}
+	if !(bytes.Equal(sha256Pre, sig[0:19])) {
+		t.Fatal("RSA signature doesn't start with prefix")
+	}
+
+	pub := ps.Public()
+	// Check public key is of right type
+	_ = pub.(*rsa.PublicKey)
+
+	ps = setup(t, "ec")
+	sig = sign(t, ps)
+
+	if !(bytes.Equal(signInput, sig)) {
+		t.Fatal("ECDSA signature error")
+	}
+
+	pub = ps.Public()
+	// Check public key is of right type
+	ecPub := pub.(*ecdsa.PublicKey)
+	if !(bytes.Equal(ecPub.X.Bytes(), ecPoint[3:35]) &&
+		bytes.Equal(ecPub.Y.Bytes(), ecPoint[35:])) {
+		t.Fatal("Incorrect decoding of EC Point")
+	}
+
+	// Disable this test because it can only work in go 1.5 and later
+	// if strings.Compare(ecPub.Curve.Params().Name, "P-256") != 0 {
+	// 	t.Fatal("Invalid curve decoded")
+	// }
+
+	k := Key{
+		module:     mockCtx{},
+		tokenLabel: "token label",
+		pin:        "unused",
+	}
+
+	// Trying to load private EC key with no public key
+	err := k.setup("no_pub_ec")
+	if err == nil {
+		t.Fatalf("Unexpected succes")
```

Comment:
*success. Also below.

This should probably be an Errorf instead of a Fatalf, because the next test case can still run and produce meaningful output even if this one fails. That allows a single test run to produce more info.